### PR TITLE
[CALCITE-3457] RexSimplify incorrectly simplifies IS NOT NULL operator with ITEM call

### PR DESCRIPTION
**Pull request by vvysotskyi**

### CODE REVIEW

File: core/src/test/resources/org/apache/calcite/test/RelOptRulesTest.xml [hunk 1]
```diff
@@ -1990,7 +2018,8 @@ LogicalProject(EXPR$0=[+($1, $5)], EXPR$1=[SUM($6) OVER (PARTITION BY $4 RANGE B
     </TestCase>
     <TestCase name="testPushProjectPastSetOp">
         <Resource name="sql">
-            <![CDATA[select sal from (select * from emp e1 union all select * from emp e2)]]>
+            <![CDATA[select sal from
+(select * from emp e1 union all select * from emp e2)]]>
         </Resource>
```
Comment (danny0405) [inline core/src/test/resources/org/apache/calcite/test/RelOptRulesTest.xml]: These changes seem not necessary, can you revert it back ?

Comment (vvysotskyi) [reply core/src/test/resources/org/apache/calcite/test/RelOptRulesTest.xml]: I have updated this file as it was recommended in this comment: https://issues.apache.org/jira/browse/CALCITE-3393?focusedCommentId=16958082&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-16958082. But if you insist, I can leave only changes required for my tests.

Comment (danny0405) [reply core/src/test/resources/org/apache/calcite/test/RelOptRulesTest.xml]: Thanks, let's keep it as it is now.

File: core/src/main/java/org/apache/calcite/rex/RexSimplify.java [hunk 2]
```diff
@@ -821,6 +823,49 @@ private RexNode simplifyIsNull(RexNode a) {
...
+      } else {
+        assert operands.stream()
+            .map(RexNode::getType)
+            .noneMatch(RelDataType::isNullable);
+      }
+    }
+  }
+
+  /**
+   * Returns {@code true} if specified {@link SqlKind} sql kind has custom nullability rules.
```
Comment (jinxing64) [inline core/src/main/java/org/apache/calcite/rex/RexSimplify.java]: IIUC it means that whether expression is null does not only depend on if the input operands contains null, but also depend on its own calculation logic.
@vvysotskyi  shall we doc and explain more here ?

Comment (vvysotskyi) [reply core/src/main/java/org/apache/calcite/rex/RexSimplify.java]: Yes, thats correct. I have corrected JavaDoc and provided an example with CAST there.

Comment (jinxing64) [reply core/src/main/java/org/apache/calcite/rex/RexSimplify.java]: Thanks a lot ~

**Pull request closed without merging**